lib/utils/envutils/environment.go

```diff
@@ -104,35 +118,48 @@ func (e *SafeEnv) Add(k, v string) {
 	v = strings.TrimSpace(v)
 	if k == "" || k == "=" {
 		return
-	}
-
-	for _, unsafeKey := range unsafeEnvironmentVars {
-		if strings.EqualFold(k, unsafeKey) {
-			return
-		}
+	} else if e.unsafeKey(k) {
+		return
 	}
 
 	*e = append(*e, fmt.Sprintf("%s=%s", k, v))
 }
 
-// AddFull adds an exact value, typically in KEY=VALUE format.  This should only be used if they values are already
-// combined.
+// AddFull adds an exact value, in the KEY=VALUE format. This should only be used if they values are already
+// combined.  When the values are separate the [Add] function is generally preferred.
 func (e *SafeEnv) AddFull(fullValues ...string) {
 valueLoop:
 	for _, kv := range fullValues {
 		kv = strings.TrimSpace(kv)
 
-		for _, unsafeKey := range unsafeEnvironmentVars {
-			if strings.HasPrefix(strings.ToUpper(kv), unsafeKey) {
-				continue valueLoop
-			}
+		key := strings.SplitN(kv, "=", 2)[0]
+		if key == "" { // weird case if the string is empty or '='
+			continue valueLoop
+		} else if e.unsafeKey(key) {
+			continue valueLoop
 		}
 
 		*e = append(*e, kv)
 	}
 }
 
-// AddExecEnvironment will add safe values from [os.Environ].
+func (e *SafeEnv) unsafeKey(key string) bool {
+	upperKey := strings.ToUpper(key)
+	if _, unsafe := unsafeEnvironmentVars[upperKey]; unsafe {
+		return true
+	}
+
+	prefix := upperKey + "="
+	for _, kv := range *e {
+		if strings.HasPrefix(strings.ToUpper(kv), prefix) {
```

Comment:
They are in linux and macOS, but they are not in windows.

Reply:
Actually this check is mostly to protect Windows, arguably we could be breaking Linux / MacOS.  In Windows regardless of the casing used for lookup you would find a value.  However in Linux and MacOS the lookup must match the casing exactly.  So, if some tooling needs to use the key `Foo` and `FOO` as two separate and distinct values, then we would not allow that distinction.

That said, I feel like this condition is extremely corner case.  And I think it's reasonable for `Teleport` to generally not allow any duplication regardless of casing.  But if necessary we can put OS specific logic in to handle the way different operating systems consider environment key conflicts.

Reply:
I would be extremely surprised if people are passing different values through a variation in casing.  For that reason, I feel that considering duplicates generally to be case insensitive is fine for simplicity of code and generally being defensive.  However if we need to do OS specific logic I can add that.

@zmb3 and @r0mant what are your thoughts?

Reply:
I think it would be least disruptive if we could do the following:

- Still allow both `foo` and `Foo` in the PAM config, maintaining existing behavior.
- Prevent overriding any variation of an existing variable in `.tsh/environment`. This means if `Foo` is already set, then any variation (`foo`, `FoO`, `FOO`, etc) of that name in `.tsh/environment` is ignored.

Seem reasonble?

Reply:
Sounds good to me, I made this change in https://github.com/gravitational/teleport/pull/34277/commits/dbc58963729d52a3b27e9efce017b62c7cb96c50